File: 18. 4Sum/Java/4Sum.java
public class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        for (int i = 0; i < nums.length - 3; i++) {
            if (i == 0 || nums[i] > nums[i - 1]) {
                for (int j = i + 1; j < nums.length - 2; j++) {
                    if (j == i + 1 || nums[j] > nums[j - 1]) {
                        int low = j + 1;
                        int high = nums.length - 1;
                        int goal = target - nums[i] - nums[j];
                        while (low < high) {
                            if (nums[low] + nums[high] == goal) {
                                result.add(Arrays.asList(nums[i], nums[j], nums[low], nums[high]));
                                while (low < high && nums[low] == nums[low + 1]) {
                                    low++;
                                }
                                low++;
                                while (low < high && nums[high] == nums[high - 1]) {
                                    high--;
                                }
                                high--;
                            } else if (nums[low] + nums[high] > goal) {
                                high--;
                            } else {
                                low++;
                            }
                        }
                    }
                }
            }
        }
        return result;
    }
}

File: 018. 4Sum/Python/4Sum.py
class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        nums.sort()
        result = []
        for i in range(len(nums) - 3):
            if i == 0 or nums[i] > nums[i - 1]:
                for j in range(i + 1, len(nums) - 2):
                    if j == i + 1 or nums[j] > nums[j - 1]:
                        low = j + 1
                        high = len(nums) - 1
                        goal = target - nums[i] - nums[j]
                        while low < high:
                            if nums[low] + nums[high] == goal:
                                result.append([nums[i], nums[j], nums[low], nums[high]])
                                while low < high and nums[low] == nums[low + 1]:
                                    low += 1
                                while low < high and nums[high] == nums[high - 1]:
                                    high -= 1
                                high -= 1
                            elif nums[low] + nums[high] > goal:
                                high -= 1
                            else:
                                low += 1
        return result
